node/network/src/peer_manager.rs

```diff
@@ -142,16 +148,36 @@ impl<T: ChainStateRetriever> PeerManager<T> {
                     future::ok(())
                 })
                 .map(|_| ())
-                .map_err(|e| warn!(target: "network", "Error processing incoming connection {}", e));
+                .map_err(
+                    |e| warn!(target: "network", "Error processing incoming connection {}", e),
+                );
             tokio::spawn(task);
         }
 
         Self { all_peer_states, node_info, phantom: PhantomData }
     }
 
     /// Ban this peer.
-    pub fn ban_peer(&self, _peer_id: &PeerId) {
-        // TODO(??): disconnect from this peer and don't accept connections.
+    pub fn ban_peer(&self, peer_id: &PeerId, duration: Duration) {
+        if let Some(state) = self.all_peer_states.write().expect(POISONED_LOCK_ERR).get_mut(peer_id)
+        {
+            let mut guard = state.write().expect(POISONED_LOCK_ERR);
+            match guard.deref() {
+                PeerState::Connected { info, .. }
+                | PeerState::Connecting { info, .. }
+                | PeerState::Ready { info, .. }
+                | PeerState::Unconnected { info, .. } => {
+                    *guard = PeerState::Unconnected {
+                        info: info.clone(),
+                        connect_timer: get_delay(CONNECT_TIMEOUT),
+                        banned_until: Some(Instant::now() + duration),
+                        evicted: false,
+                    };
+                }
+                // Incoming connection will be terminated when it is processed.
+                PeerState::IncomingConnection { .. } => (),
```

Comment:
I don't think it will. The code that this PR modifies in `IncommingConnection` is about dropping it when there is already another peer instance that was created and banned.

Reply:
Yes. But if you connect with the same peer_id, it is considered as the same instance because `all_peer_states` is keyed by `PeerId`. If you connect with different id, the ban cannot be triggered for incoming connections.

Reply:
We should explicitly panic here then. Also, add into docstring section on panics. See how to format docstring here: https://doc.rust-lang.org/src/core/sync/atomic.rs.html#328

Reply:
Fixed